contracts/actions/mcd/McdBoostComposite.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import "../ActionBase.sol";
+import "../exchange/DFSSell.sol";
+import "../fee/GasFeeTaker.sol";
+
+import "../../interfaces/mcd/IDaiJoin.sol";
+import "../../interfaces/mcd/IJug.sol";
+
+import "./helpers/McdHelper.sol";
+import "./helpers/McdRatioHelper.sol";
+
+/// @title Single mcd boost action can use flashloan or not
+contract McdBoostComposite is ActionBase, DFSSell, GasFeeTaker, McdHelper, McdRatioHelper {
+    using TokenUtils for address;
+
+    error RatioNotLowerThanBefore(uint256, uint256);
+    error WrongAsset(address);
+    error TargetRatioMiss(uint256, uint256);
+
+    /// @dev 2% offset acceptable
+    uint256 internal constant RATIO_OFFSET = 20000000000000000;
+
+    /// @param vaultId Id of the vault
+    /// @param joinAddr Collateral join address
+    /// @param gasUsed Gas amount to charge in strategies
+    /// @param flAddr Flashloan address 0x0 if we're not using flashloan
+    /// @param flAmount Amount that the flashloan actions returns if used (must have it because of fee)
+    /// @param nextPrice Maker OSM next price if 0 we're using current price (used for ratio check)
+    /// @param targetRatio Target ratio to repay if 0 we are not checking the ratio
+    /// @param exchangeData Data needed for swap
+    struct BoostParams {
+        uint256 vaultId;
+        address joinAddr;
+        uint256 gasUsed;
+        address flAddr;
+        uint256 flAmount;
+        uint256 nextPrice;
+        uint256 targetRatio;
+        ExchangeData exchangeData;
+    }
+
+    /// @inheritdoc ActionBase
+    function executeAction(
+        bytes calldata _callData,
+        bytes32[] memory _subData,
+        uint8[] memory _paramMapping,
+        bytes32[] memory _returnValues
+    ) public payable virtual override(ActionBase, DFSSell, GasFeeTaker) returns (bytes32) {
+        BoostParams memory boostParams = _parseCompositeParams(_callData);
+
+        boostParams.vaultId = _parseParamUint(
+            boostParams.vaultId,
+            _paramMapping[0],
+            _subData,
+            _returnValues
+        );
+
+        boostParams.joinAddr = _parseParamAddr(
+            boostParams.joinAddr,
+            _paramMapping[1],
+            _subData,
+            _returnValues
+        );
+
+        boostParams.flAmount = _parseParamUint(
+            boostParams.flAmount,
+            _paramMapping[2],
+            _subData,
+            _returnValues
+        );
+
+        boostParams.exchangeData.srcAddr = _parseParamAddr(
```

Comment:
seems redundant to map this as we already have a check below forcing this to be DAI

Reply:
similar for RepayComposite